realm/realm-library/src/main/cpp/io_realm_SyncSession.cpp

```diff
@@ -49,3 +50,67 @@ JNIEXPORT jboolean JNICALL Java_io_realm_SyncSession_nativeRefreshAccessToken(JN
     CATCH_STD()
     return JNI_FALSE;
 }
+
+
+JNIEXPORT jlong JNICALL Java_io_realm_SyncSession_nativeAddProgressListener(JNIEnv* env, jclass,
+                                                                            jstring j_local_realm_path,
+                                                                            jlong listener_id, jint direction,
+                                                                            jboolean is_streaming)
+{
+    try {
+        // JNIEnv is thread confined, so we need a deep copy in order to capture the string in the lambda
+        std::string local_realm_path(JStringAccessor(env, j_local_realm_path));
+        std::shared_ptr<SyncSession> session = SyncManager::shared().get_existing_active_session(local_realm_path);
+        if (!session) {
+            // FIXME: We should lift this restriction
+            ThrowException(env, IllegalState,
+                           "Cannot register a progress listener before a session is "
+                           "created. A session will be created after the first call to Realm.getInstance().");
+            return static_cast<jlong>(0);
+        }
+
+        SyncSession::NotifierType type =
+            (direction == 1) ? SyncSession::NotifierType::download : SyncSession::NotifierType::upload;
+
+        std::function<SyncProgressNotifierCallback> callback = [local_realm_path, listener_id](
+            uint64_t transferred, uint64_t transferrable) {
+            JNIEnv* local_env = jni_util::JniUtils::get_env(true);
+
+            auto path = local_env->NewStringUTF(local_realm_path.c_str());
+            local_env->CallStaticVoidMethod(java_syncmanager_class, java_notify_progress_listener, path, listener_id,
+                                            static_cast<jlong>(transferred), static_cast<jlong>(transferrable));
+
+            // All exceptions will be caught on the Java side of handlers, but errors will still end
+            // up here, so we need to do something sensible with them.
+            // Throwing a C++ exception will terminate the sync thread and cause the pending Java
+            // exception to become visible. For some (unknown) reason Logcat will not see the C++
+            // exception, only the Java one.
+            if (local_env->ExceptionCheck()) {
+                local_env->ExceptionDescribe(local_env);
+                throw std::runtime_error("An unexpected Error was thrown from Java. See LogCat");
+            }
+
+            // Callback happens on a thread not controlled by the JVM. So manual cleanup is
+            // required.
```

Comment:
I feel this is still needed, since local ref will be released when call from JNI back to java. But case here is like:
1 C++ (acquire local ref) -> 2 java -> C++

I am not quite sure the local ref will be automatically released when cross the 1->2 border since the bottom call is not in java like like our normal case:
1 Java -> 2 C++ (acquire local ref) -> 3 back to java